File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Solution {
	public String[] solution(String[] record) {
		Map<String, String> users = new HashMap<>();
		List<String[]> commandList = new ArrayList<>();

		for (String log : record) {
			String[] commands = log.split(" ");
			switch (commands[0]) {
				case "Enter":
				case "Change":
					users.put(commands[1], commands[2]);
					break;
				default:
					break;
			}
			commandList.add(commands);
		}

		List<String> answer = new ArrayList<>();

		for (String[] commands : commandList) {
			switch (commands[0]) {
				case "Enter":
					answer.add(users.get(commands[1]) + "님이 들어왔습니다.");
					break;
				case "Leave":
					answer.add(users.get(commands[1]) + "님이 나갔습니다.");
					break;
				default:
					break;
			}
		}

		return answer.toArray(String[]::new);
	}
}

File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.py
from typing import List


def solution(record: List[str]) -> List[str]:
    user_map = {}
    system_logs: List[List[str]] = []
    
    for i in record:
        strings = i.split(" ")
        
        if strings[0] == "Enter":
            user_map[strings[1]] = strings[2]
            system_logs.append(strings[:2])
        elif strings[0] == "Leave":
            system_logs.append(strings)
        elif strings[0] == "Change":
            user_map[strings[1]] = strings[2]
            
    final_logs = []
    
    for log in system_logs:
        if log[0] == "Enter":
            final_logs.append(f"{user_map[log[1]]}님이 들어왔습니다.")
        else:
            final_logs.append(f"{user_map[log[1]]}님이 나갔습니다.")
            
    return final_logs